File: ld_estimator/flip.py

from ld_estimator.dprime import get_d

def fliphap(hap):
    ''' flip frequencies in a Haps object
    '''
    hap.aa, hap.ab = hap.ab, hap.aa
    hap.ba, hap.bb = hap.bb, hap.ba
    return hap

def flip_freqs(freqs, known, pA2, pB2):
    ''' flip frequencies to get positive D'
    '''
    # flip AA with AB and BA with BB
    freqs = fliphap(freqs)

    # flip frequency of second allele
    pA2, pB2 = pB2, pA2

    return freqs, fliphap(known), pA2, pB2, get_d(freqs)

File: src/flip.cpp

#include "flip.h"

namespace ld_estimator {

template <typename T>
void fliphap(Haps<T>& hap) {
  // flip AA with AB and BA with BB
  T aa = hap.aa;
  T ab = hap.ab;
  T ba = hap.ba;
  T bb = hap.bb;

  // flip the values
  hap.ab = aa;
  hap.aa = ab;
  hap.ba = bb;
  hap.bb = ba;
}

void flip_freqs(Haps<double>& freqs, Haps<int>& known, double& pA2, double& pB2, double& d) {
  //flip frequencies to get positive D'
  fliphap(freqs);
  fliphap(known);
  d = get_d(freqs);

  // flip frequency of second allele
  std::swap(pA2, pB2);
}

}
